File: MiniDOMBase/DMD/Anagrafica/Ricontatti/CRMFilter.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases



Partial Public Class Anagrafica

    <Flags> _
    Public Enum CRMFilterFlags As Integer
        NONE = 0
        'APPUNTAMENTI = 1
        'TELEFONATE = 2
        SOLOAZIENDE = 4

        SOLOINCORSO = 8
    End Enum

    Public Enum CRMFilterSortFlags As Integer
        MOSTIMPORTANT = 0
        MOSTRECENT = 1
        LEASTRECENT = 2
        NAME = 3
    End Enum

    Public Class CRMFilter
        Implements XML.IDMDXMLSerializable, ICloneable, IComparer '(Of CActivePerson)

        Public TipiAppuntamento As New CCollection(Of String)
        Public Categorie As New CCollection(Of String)
        Public TipiRapporto As New CCollection(Of String)
        Public Flags As CRMFilterFlags = CRMFilterFlags.NONE ' CRMFilterFlags.APPUNTAMENTI Or CRMFilterFlags.TELEFONATE
        Public Motivo As String = ""
        Public NomeLista As String = ""
        Public IDPuntoOperativo As Integer = 0
        Public Periodo As String = ""
        Public DataInizio As Date? = Nothing
        Public DataFine As Date? = Nothing
        Public IDOperatore As Integer = 0
        Public SortOrder As CRMFilterSortFlags = CRMFilterSortFlags.MOSTIMPORTANT
        Public nMax As Integer? = Nothing
        Public fromDate As Date? = Nothing
        Public ResidenteA As String = ""

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Public Property MostraSoloAziende As Boolean
            Get
                Return Sistema.TestFlag(Me.Flags, CRMFilterFlags.SOLOAZIENDE)
            End Get
            Set(value As Boolean)
                Me.Flags = Sistema.SetFlag(Me.Flags, CRMFilterFlags.SOLOAZIENDE, value)
            End Set
        End Property

        Public Property MostraTelefonate As Boolean
            Get
                Return (Me.TipiAppuntamento.Count() = 0) OrElse (Me.TipiAppuntamento.Contains("Telefonata"))
            End Get
            Set(value As Boolean)
                If (value) Then
                    If (Not Me.TipiAppuntamento.Contains("Telefonata")) Then Me.TipiAppuntamento.Add("Telefonata")
                Else
                    If (Me.TipiAppuntamento.Contains("Telefonata")) Then Me.TipiAppuntamento.Remove("Telefonata")
                End If
            End Set
        End Property

        Public Property MostraAppuntamenti As Boolean
            Get
                Return (Me.TipiAppuntamento.Count() = 0) OrElse (Me.TipiAppuntamento.Contains("Appuntamento"))
            End Get
            Set(value As Boolean)
                If (value) Then
                    If (Not Me.TipiAppuntamento.Contains("Appuntamento")) Then Me.TipiAppuntamento.Add("Appuntamento")
                Else
                    If (Me.TipiAppuntamento.Contains("Appuntamento")) Then Me.TipiAppuntamento.Remove("Appuntamento")
                End If
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta un valore booleano che indica se la ricerca deve restituire o meno i promemoria
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property MostraPromemoria As Boolean
            Get
                Return (Me.TipiAppuntamento.Count() = 0) OrElse (Me.TipiAppuntamento.Contains("Promemoria"))
            End Get
            Set(value As Boolean)
                If (value) Then
                    If (Not Me.TipiAppuntamento.Contains("Promemoria")) Then Me.TipiAppuntamento.Add("Promemoria")
                Else
                    If (Me.TipiAppuntamento.Contains("Promemoria")) Then Me.TipiAppuntamento.Remove("Promemoria")
                End If
            End Set
        End Property

        Public Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "TipiAppuntamento" : Me.TipiAppuntamento.Clear() : Me.TipiAppuntamento.AddRange(fieldValue)
                Case "Categorie" : Me.Categorie.Clear() : Me.Categorie.AddRange(fieldValue)
                Case "TipiRapporto" : Me.TipiRapporto.Clear() : Me.TipiRapporto.AddRange(fieldValue)
                Case "Flags" : Me.Flags = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Motivo" : Me.Motivo = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "NomeLista" : Me.NomeLista = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "IDPuntoOperativo" : Me.IDPuntoOperativo = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Periodo" : Me.Periodo = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "DataInizio" : Me.DataInizio = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "DataFine" : Me.DataFine = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "IDOperatore" : Me.IDOperatore = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "SortOrder" : Me.SortOrder = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "nMax" : Me.nMax = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "fromDate" : Me.fromDate = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "ResidenteA" : Me.ResidenteA = XML.Utils.Serializer.DeserializeString(fieldValue)
            End Select
        End Sub

        Public Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("Flags", Me.Flags)
            writer.WriteAttribute("Motivo", Me.Motivo)
            writer.WriteAttribute("NomeLista", Me.NomeLista)
            writer.WriteAttribute("IDPuntoOperativo", Me.IDPuntoOperativo)
            writer.WriteAttribute("Periodo", Me.Periodo)
            writer.WriteAttribute("DataInizio", Me.DataInizio)
            writer.WriteAttribute("DataFine", Me.DataFine)
            writer.WriteAttribute("IDOperatore", Me.IDOperatore)
            writer.WriteAttribute("SortOrder", Me.SortOrder)
            writer.WriteAttribute("nMax", Me.nMax)
            writer.WriteAttribute("fromDate", Me.fromDate)
            writer.WriteAttribute("ResidenteA", Me.ResidenteA)
            writer.WriteTag("TipiAppuntamento", Me.TipiAppuntamento)
            writer.WriteTag("Categorie", Me.Categorie)
            writer.WriteTag("TipiRapporto", Me.TipiRapporto)
        End Sub

        Public Function check(ByVal ric As CRicontatto) As Boolean
            Dim p As CPersona = ric.Persona()


            '(ric.NomeLista = Me.NomeLista) AndAlso
            Dim ret As Boolean =
                   (ric.IDPuntoOperativo = Me.IDPuntoOperativo OrElse ric.IDAssegnatoA = Me.IDOperatore) AndAlso
                   DateUtils.CheckBetween(ric.DataPrevista, Me.DataInizio, Me.DataFine) AndAlso
                   (Me.Motivo = "" OrElse Strings.InStr(ric.Note, Me.Motivo) > 0)
            If (ret AndAlso Me.TipiAppuntamento.Count() > 0) Then
                ret = False
                For Each value As String In Me.TipiAppuntamento
                    If (ric.TipoAppuntamento = value) Then
                        ret = True
                        Exit For
                    End If
                Next
            End If
            If (ret AndAlso Me.TipiRapporto.Count() > 0) Then
                p = ric.Persona()
                Dim tr As String = ""
                If ((TypeOf (p) Is CPersonaFisica) AndAlso DirectCast(p, CPersonaFisica).ImpiegoPrincipale() IsNot Nothing) Then
                    tr = DirectCast(p, CPersonaFisica).ImpiegoPrincipale.TipoRapporto
                End If
                For Each value As String In Me.TipiRapporto
                    If (value = tr) Then
                        ret = True
                        Exit For
                    End If
                Next
            End If
            If (ret AndAlso Me.Categorie.Count() > 0) Then
                ret = False
                For Each value As String In Me.Categorie
                    If (ric.Categoria = value) Then
                        ret = True
                        Exit For
                    End If
                Next
            End If
            If (ret AndAlso Me.ResidenteA <> "") Then
                p = ric.Persona
                ret = ret AndAlso p.ResidenteA.NomeComune = Me.ResidenteA
            End If
            Return ret
        End Function

        Public Function Clone() As Object Implements ICloneable.Clone
            Dim ret As New CRMFilter()
            ret.TipiAppuntamento.AddRange(Me.TipiAppuntamento)
            ret.Categorie.AddRange(Me.Categorie)
            ret.TipiRapporto.AddRange(Me.TipiRapporto)
            ret.Flags = Me.Flags
            ret.Motivo = Me.Motivo
            ret.NomeLista = Me.NomeLista
            ret.IDPuntoOperativo = Me.IDPuntoOperativo
            ret.DataInizio = Me.DataInizio
            ret.DataFine = Me.DataFine
            ret.IDOperatore = Me.IDOperatore
            ret.Periodo = Me.Periodo
            ret.SortOrder = Me.SortOrder
            ret.nMax = Me.nMax
            ret.fromDate = Me.fromDate
            ret.ResidenteA = Me.ResidenteA
            Return ret
        End Function

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub

        Private Function Compare1(x As Object, y As Object) As Integer Implements IComparer.Compare
            Return Me.Compare(x, y)
        End Function

        Public Function Compare(ByVal a As CActivePerson, b As CActivePerson) As Integer
            Dim ret As Integer = 0
            Select Case Me.SortOrder
                Case CRMFilterSortFlags.LEASTRECENT
                    ret = DateUtils.Compare(a.Data, b.Data)
                    If (ret = 0) Then ret = Strings.Compare(a.Nominativo, b.Nominativo, CompareMethod.Text)
                Case CRMFilterSortFlags.MOSTIMPORTANT
                    Dim p1 As Integer = 0
                    Dim p2 As Integer = 0
                    If (a.Ricontatto IsNot Nothing) Then p1 = a.Ricontatto.Priorita
                    If (b.Ricontatto IsNot Nothing) Then p2 = b.Ricontatto.Priorita
                    ret = p1.CompareTo(p2)
                Case CRMFilterSortFlags.MOSTRECENT
                    ret = -DateUtils.Compare(a.Data, b.Data)
                    If (ret = 0) Then ret = Strings.Compare(a.Nominativo, b.Nominativo, CompareMethod.Text)
                Case CRMFilterSortFlags.NAME
                    ret = Strings.Compare(a.Nominativo, b.Nominativo, CompareMethod.Text)
                Case Else
                    ret = -DateUtils.Compare(a.Data, b.Data)
                    If (ret = 0) Then ret = Strings.Compare(a.Nominativo, b.Nominativo, CompareMethod.Text)
            End Select
            Return ret
        End Function
    End Class


End Class

File: MiniDOMBase/DMD/Anagrafica/Ricontatti/Providers/CRMFilter.cs
using System;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using System.Collections;
using System.Collections.Generic;

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Flag per il filtro sui ricontatti
        /// </summary>
        [Flags]
        public enum CRMFilterFlags : int
        {
            /// <summary>
            /// Nessun flag
            /// </summary>
            NONE = 0,
            // APPUNTAMENTI = 1
            // TELEFONATE = 2

            /// <summary>
            /// Restituisce solo le azienda
            /// </summary>
            SOLOAZIENDE = 4,

            /// <summary>
            /// Restituisce solo i clienti attivi
            /// </summary>
            SOLOINCORSO = 8
        }

        /// <summary>
        /// Tipi di ordinamento dei risultati
        /// </summary>
        public enum CRMFilterSortFlags : int
        {
            /// <summary>
            /// Mostra prima i più importanti
            /// </summary>
            MOSTIMPORTANT = 0,

            /// <summary>
            /// Mostra prima i più recenti
            /// </summary>
            MOSTRECENT = 1,

            /// <summary>
            /// Mostra prima i meno recenti
            /// </summary>
            LEASTRECENT = 2,

            /// <summary>
            /// Ordina per nome del cliente
            /// </summary>
            NAME = 3
        }

        /// <summary>
        /// Filtro di ricerca
        /// </summary>
        public class CRMFilter 
            : IDMDXMLSerializable, ICloneable, IComparer , IComparer<CActivePerson>
        {

            /// <summary>
            /// Tipi di appuntamento da restituire
            /// </summary>
            public CCollection<string> TipiAppuntamento = new CCollection<string>();

            /// <summary>
            /// Categorie di appuntamento da restituire
            /// </summary>
            public CCollection<string> Categorie = new CCollection<string>();

            /// <summary>
            /// Tipi rapporto da restituire
            /// </summary>
            public CCollection<string> TipiRapporto = new CCollection<string>();

            /// <summary>
            /// Flags
            /// </summary>
            public CRMFilterFlags Flags = CRMFilterFlags.NONE; // CRMFilterFlags.APPUNTAMENTI Or CRMFilterFlags.TELEFONATE

            /// <summary>
            /// Motivo del ricontatto
            /// </summary>
            public string Motivo = "";

            /// <summary>
            /// Nome della lista
            /// </summary>
            public string NomeLista = "";

            /// <summary>
            /// Id del punto operativo di assegnazione
            /// </summary>
            public int IDPuntoOperativo = 0;

            /// <summary>
            /// Tipo periodo (es oggi, domani, tra)
            /// </summary>
            public string Periodo = "";

            /// <summary>
            /// Data inizio
            /// </summary>
            public DateTime? DataInizio = default;

            /// <summary>
            /// Data fine
            /// </summary>
            public DateTime? DataFine = default;

            /// <summary>
            /// Di operatore di assegnazione
            /// </summary>
            public int IDOperatore = 0;

            /// <summary>
            /// Ordinamento dei risultati
            /// </summary>
            public CRMFilterSortFlags SortOrder = CRMFilterSortFlags.MOSTIMPORTANT;

            /// <summary>
            /// Numero massimo dei risultati da restituire
            /// </summary>
            public int? nMax = default;

            /// <summary>
            /// Inizia a restituire dalla data
            /// </summary>
            public DateTime? fromDate = default;

            /// <summary>
            /// Indirizzo di residenza
            /// </summary>
            public string ResidenteA = "";

            /// <summary>
            /// Costruttore
            /// </summary>
            public CRMFilter()
            {
                ////DMDObject.IncreaseCounter(this);
            }

            /// <summary>
            /// Mostra solo le persone giuridiche
            /// </summary>
            public bool MostraSoloAziende
            {
                get
                {
                    return DMD.RunTime.TestFlag(Flags, CRMFilterFlags.SOLOAZIENDE);
                }

                set
                {
                    Flags = DMD.RunTime.SetFlag(Flags, CRMFilterFlags.SOLOAZIENDE, value);
                }
            }

            /// <summary>
            /// Mostra le telefonate
            /// </summary>
            public bool MostraTelefonate
            {
                get
                {
                    return TipiAppuntamento.Count == 0 || DMD.Booleans.ValueOf(TipiAppuntamento.Contains("Telefonata"));
                }

                set
                {
                    if (value)
                    {
                        if (!TipiAppuntamento.Contains("Telefonata"))
                            TipiAppuntamento.Add("Telefonata");
                    }
                    else if (DMD.Booleans.ValueOf(TipiAppuntamento.Contains("Telefonata")))
                        TipiAppuntamento.Remove("Telefonata");
                }
            }

            /// <summary>
            /// Mostra gli appuntamenti
            /// </summary>
            public bool MostraAppuntamenti
            {
                get
                {
                    return TipiAppuntamento.Count == 0 || DMD.Booleans.ValueOf(TipiAppuntamento.Contains("Appuntamento"));
                }

                set
                {
                    if (value)
                    {
                        if (!TipiAppuntamento.Contains("Appuntamento"))
                            TipiAppuntamento.Add("Appuntamento");
                    }
                    else if (DMD.Booleans.ValueOf(TipiAppuntamento.Contains("Appuntamento")))
                        TipiAppuntamento.Remove("Appuntamento");
                }
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return HashCalculator.Calculate(this.Motivo, this.NomeLista, this.Periodo);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public sealed override bool Equals(object obj)
            {
                return (obj is CRMFilter) && this.Equals((CRMFilter)obj);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(CRMFilter obj )
            {
                return
                       CollectionUtils.EQAnyOrder(this.TipiAppuntamento, obj.TipiAppuntamento)
                    && CollectionUtils.EQAnyOrder(this.Categorie, obj.Categorie)
                    && CollectionUtils.EQAnyOrder(this.TipiRapporto, obj.TipiRapporto)
                    && DMD.Integers.EQ((int)this.Flags, (int)obj.Flags)
                    && DMD.Strings.EQ(this.Motivo, obj.Motivo)
                    && DMD.Strings.EQ(this.NomeLista, obj.NomeLista)
                    && DMD.Integers.EQ(this.IDPuntoOperativo, obj.IDPuntoOperativo)
                    && DMD.Strings.EQ(this.Periodo, obj.Periodo)
                    && DMD.DateUtils.EQ(this.DataInizio, obj.DataInizio)
                    && DMD.DateUtils.EQ(this.DataFine, obj.DataFine)
                    && DMD.Integers.EQ(this.IDOperatore, obj.IDOperatore)
                    && DMD.Integers.EQ((int)this.SortOrder, (int)obj.SortOrder)
                    && DMD.Integers.EQ(this.nMax, obj.nMax)
                    && DMD.DateUtils.EQ(this.fromDate, obj.fromDate)
                    && DMD.Strings.EQ(this.ResidenteA, obj.ResidenteA)
             ;

        }

        /// <summary>
        /// Restituisce o imposta un valore booleano che indica se la ricerca deve restituire o meno i promemoria
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public bool MostraPromemoria
            {
                get
                {
                    return TipiAppuntamento.Count == 0 || DMD.Booleans.ValueOf(TipiAppuntamento.Contains("Promemoria"));
                }

                set
                {
                    if (value)
                    {
                        if (!TipiAppuntamento.Contains("Promemoria"))
                            TipiAppuntamento.Add("Promemoria");
                    }
                    else if (DMD.Booleans.ValueOf(TipiAppuntamento.Contains("Promemoria")))
                        TipiAppuntamento.Remove("Promemoria");
                }
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            public void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "TipiAppuntamento":
                        {
                            TipiAppuntamento.Clear();
                            TipiAppuntamento.AddRange((IEnumerable)fieldValue);
                            break;
                        }

                    case "Categorie":
                        {
                            Categorie.Clear();
                            Categorie.AddRange((IEnumerable)fieldValue);
                            break;
                        }

                    case "TipiRapporto":
                        {
                            TipiRapporto.Clear();
                            TipiRapporto.AddRange((IEnumerable)fieldValue);
                            break;
                        }

                    case "Flags":
                        {
                            Flags = (CRMFilterFlags)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Motivo":
                        {
                            Motivo = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "NomeLista":
                        {
                            NomeLista = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "IDPuntoOperativo":
                        {
                            IDPuntoOperativo = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Periodo":
                        {
                            Periodo = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "DataInizio":
                        {
                            DataInizio = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "DataFine":
                        {
                            DataFine = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "IDOperatore":
                        {
                            IDOperatore = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "SortOrder":
                        {
                            SortOrder = (CRMFilterSortFlags)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "nMax":
                        {
                            nMax = DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "fromDate":
                        {
                            fromDate = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "ResidenteA":
                        {
                            ResidenteA = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            public void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Flags", (int?)Flags);
                writer.WriteAttribute("Motivo", Motivo);
                writer.WriteAttribute("NomeLista", NomeLista);
                writer.WriteAttribute("IDPuntoOperativo", IDPuntoOperativo);
                writer.WriteAttribute("Periodo", Periodo);
                writer.WriteAttribute("DataInizio", DataInizio);
                writer.WriteAttribute("DataFine", DataFine);
                writer.WriteAttribute("IDOperatore", IDOperatore);
                writer.WriteAttribute("SortOrder", (int?)SortOrder);
                writer.WriteAttribute("nMax", nMax);
                writer.WriteAttribute("fromDate", fromDate);
                writer.WriteAttribute("ResidenteA", ResidenteA);
                writer.WriteTag("TipiAppuntamento", TipiAppuntamento);
                writer.WriteTag("Categorie", Categorie);
                writer.WriteTag("TipiRapporto", TipiRapporto);
            }

            /// <summary>
            /// Verifica se il ricontatto verifica i requisiti del filtro
            /// </summary>
            /// <param name="ric"></param>
            /// <returns></returns>
            public bool check(CRicontatto ric)
            {
                var p = ric.Persona;


                // (ric.NomeLista = Me.NomeLista) AndAlso
                bool ret = (ric.IDPuntoOperativo == IDPuntoOperativo || ric.IDAssegnatoA == IDOperatore) && DMD.DateUtils.CheckBetween(ric.DataPrevista, DataInizio, DataFine) && (string.IsNullOrEmpty(Motivo) || DMD.Strings.InStr(ric.Note, Motivo) > 0);
                if (ret && TipiAppuntamento.Count > 0)
                {
                    ret = false;
                    foreach (string value in TipiAppuntamento)
                    {
                        if ((ric.TipoAppuntamento ?? "") == (value ?? ""))
                        {
                            ret = true;
                            break;
                        }
                    }
                }

                if (ret && TipiRapporto.Count > 0)
                {
                    p = ric.Persona;
                    string tr = "";
                    if (p is CPersonaFisica && ((CPersonaFisica)p).ImpiegoPrincipale is object)
                    {
                        tr = ((CPersonaFisica)p).ImpiegoPrincipale.TipoRapporto;
                    }

                    foreach (string value in TipiRapporto)
                    {
                        if ((value ?? "") == (tr ?? ""))
                        {
                            ret = true;
                            break;
                        }
                    }
                }

                if (ret && Categorie.Count > 0)
                {
                    ret = false;
                    foreach (string value in Categorie)
                    {
                        if ((ric.Categoria ?? "") == (value ?? ""))
                        {
                            ret = true;
                            break;
                        }
                    }
                }

                if (ret && !string.IsNullOrEmpty(ResidenteA))
                {
                    p = ric.Persona;
                    ret = ret && (p.ResidenteA.NomeComune ?? "") == (ResidenteA ?? "");
                }

                return ret;
            }

            /// <summary>
            /// Clona il filtro
            /// </summary>
            /// <returns></returns>
            public object Clone()
            {
                var ret = new CRMFilter();
                ret.TipiAppuntamento.AddRange(TipiAppuntamento);
                ret.Categorie.AddRange(Categorie);
                ret.TipiRapporto.AddRange(TipiRapporto);
                ret.Flags = Flags;
                ret.Motivo = Motivo;
                ret.NomeLista = NomeLista;
                ret.IDPuntoOperativo = IDPuntoOperativo;
                ret.DataInizio = DataInizio;
                ret.DataFine = DataFine;
                ret.IDOperatore = IDOperatore;
                ret.Periodo = Periodo;
                ret.SortOrder = SortOrder;
                ret.nMax = nMax;
                ret.fromDate = fromDate;
                ret.ResidenteA = ResidenteA;
                return ret;
            }

            /// <summary>
            /// Distruttore
            /// </summary>
            ~CRMFilter()
            {
                // //DMDObject.DecreaseCounter(this);
            }

            int IComparer.Compare(object x, object y)
            {
                return Compare((Sistema.CActivePerson)x, (Sistema.CActivePerson)y);
            }

            /// <summary>
            /// Compara due ricontatti sulla base delle impostazioni del filtro 
            /// </summary>
            /// <param name="a"></param>
            /// <param name="b"></param>
            /// <returns></returns>
            public int Compare(Sistema.CActivePerson a, Sistema.CActivePerson b)
            {
                int ret = 0;
                switch (SortOrder)
                {
                    case CRMFilterSortFlags.LEASTRECENT:
                        {
                            ret = DMD.DateUtils.Compare(a.Data, b.Data);
                            if (ret == 0)
                                ret = DMD.Strings.Compare(a.Nominativo, b.Nominativo, true);
                            break;
                        }

                    case CRMFilterSortFlags.MOSTIMPORTANT:
                        {
                            int p1 = 0;
                            int p2 = 0;
                            if (a.Ricontatto is object)
                                p1 = a.Ricontatto.Priorita;
                            if (b.Ricontatto is object)
                                p2 = b.Ricontatto.Priorita;
                            ret = p1.CompareTo(p2);
                            break;
                        }

                    case CRMFilterSortFlags.MOSTRECENT:
                        {
                            ret = -DMD.DateUtils.Compare(a.Data, b.Data);
                            if (ret == 0)
                                ret = DMD.Strings.Compare(a.Nominativo, b.Nominativo, true);
                            break;
                        }

                    case CRMFilterSortFlags.NAME:
                        {
                            ret = DMD.Strings.Compare(a.Nominativo, b.Nominativo, true);
                            break;
                        }

                    default:
                        {
                            ret = -DMD.DateUtils.Compare(a.Data, b.Data);
                            if (ret == 0)
                                ret = DMD.Strings.Compare(a.Nominativo, b.Nominativo, true);
                            break;
                        }
                }

                return ret;
            }
        }
    }
}